Core/NoOpRunner.Core.Tests/GameStateBuilderTests.cs

```diff
@@ -0,0 +1,208 @@
+﻿using NoOpRunner.Core.Builders;
+using NoOpRunner.Core.Shapes;
+using NoOpRunner.Core.Shapes.EntityShapes;
+using NoOpRunner.Core.Shapes.GenerationStrategies;
+using NoOpRunner.Core.Shapes.StaticShapes;
+using Shouldly;
+using System;
+using System.Linq;
+using Xunit;
+
+namespace NoOpRunner.Core.Tests
+{
+    public class GameStateBuilderTests
+    {
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void GameStateBuilder_WhenMapIsUninitialized_AndShapeIsAdded_ShouldThrowException(bool initializeMap)
+        {
+            // Arrange
+            var gameStateConfigurator = new GameStateBuilder().Configure();
+
+            if (initializeMap)
+            {
+                gameStateConfigurator.InitializeMap(1, 1);
+            }
+
+            // Act
+            var action = new Action(() => gameStateConfigurator.AddShape(f => f.GetShape(Enums.Shape.Saw, 1, 1)));
+
+            // Assert
+            if (!initializeMap)
+                action.ShouldThrow<Exception>();
+            else
+                action.ShouldNotThrow();
```

Comment:
If you have a conditional statement in your assertion, that's a big red flag that you should split the test into two. Same for a few other tests. You shouldn't have conditionals in your arrange or assert stages, you pretty much shouldn't have conditionals in a test function at all. The code will be a bit more bloated when you split the tests, but easy code comprehension is what matters in tests, code repetition is fine.

Reply:
The reason for this is to make sure the same setup but without initialization clause doesn't cause an exception, but without it does. If these tests are split into separate, eventually the arrangement might become different in those two cases and the one that asserts an exception might start passing for the wrong reasons and will no longer bring any value.